bpo-18819: tarfile: only set device fields for device files

wchargin opened the pull request:
The GNU docs describe the `devmajor` and `devminor` fields of the tar
header struct only in the context of character and block special files,
suggesting that in other cases they are not populated. Typical utilities
behave accordingly; this patch teaches `tarfile` to do the same.

Test Plan:
Unit tests added; they fail before this commit and pass after it. Also
verified that this enables output that is bit-for-bit compatible with
GNU `tar`. In particular, this program now passes on my Ubuntu 16.04,
whereas it failed before this patch:

```python
import os
import subprocess
import tarfile
import tempfile

filename = "important_data"
contents = b"The quick brown fox jumps over the lazy dog"

with tempfile.TemporaryDirectory() as tmpdir:
    os.chdir(tmpdir)
    with open(filename, "wb") as outfile:
        outfile.write(contents)
    with tarfile.open("py.tar", "x", format=tarfile.GNU_FORMAT) as outfile:
        outfile.add(filename)
    subprocess.check_call(["tar", "cf", "gnu.tar", filename])
    subprocess.check_call(["cmp", "-b", "py.tar", "gnu.tar"])
```

(The exact contents of the files depend on the calling user and current
time, but should always be the same across both output archives.)

wchargin-branch: tarfile-limit-device-headers


<!-- issue-number: [bpo-18819](https://bugs.python.org/issue18819) -->
https://bugs.python.org/issue18819
<!-- /issue-number -->

wchargin committed:
bpo-18819: tarfile: only set device fields for device files

The GNU docs describe the `devmajor` and `devminor` fields of the tar
header struct only in the context of character and block special files,
suggesting that in other cases they are not populated. Typical utilities
behave accordingly; this patch teaches `tarfile` to do the same.

Test Plan:
No tests added because none appear to exist for this module. Manually
verified that this enables output that is bit-for-bit compatible with
GNU tar. In particular, this program now passes on my Ubuntu 16.04,
whereas it failed before this patch:

```python
import os
import subprocess
import tarfile
import tempfile

filename = "important_data"
contents = b"The quick brown fox jumps over the lazy dog"

with tempfile.TemporaryDirectory() as tmpdir:
    os.chdir(tmpdir)
    with open(filename, "wb") as outfile:
        outfile.write(contents)
    with tarfile.open("py.tar", "x", format=tarfile.GNU_FORMAT) as outfile:
        outfile.add(filename)
    subprocess.check_call(["tar", "cf", "gnu.tar", filename])
    subprocess.check_call(["shasum", "-a", "256", "py.tar", "gnu.tar"])
    subprocess.check_call(["cmp", "-b", "py.tar", "gnu.tar"])
```

(The exact hashes depend on the calling user and current time but should
always be the same across both output archives.)

wchargin-branch: tarfile-limit-device-headers

ethanfurman reviewed (changes requested):
This looks good.  Add your test and we should be ready to commit.

wchargin reviewed (commented):
> Question, though: why use both `cmp` and `shasum`? If `cmp` passes aren't the files identical?

Yes; I had the `shasum` just as a sanity check that the files weren’t
empty, but you’re right that the `cmp` is certainly sufficient. I’ll go
ahead and remove this from the description to avoid confusion.

> Also, the tests for `tarfile` are in `Lib/test/test_tarfile.py`.

Great, thanks. Not sure how I missed this.

> Add your test and we should be ready to commit.

Will add a test to `test_tarfile.py`, probably more like the existing
tests in that file than like the end-to-end test in the description
(which would only work on systems with GNU `tar` installed).

Thanks for the speedy review! I’ll add a test and get back to you.

wchargin committed:
review: add unit test

Test Plan:
This passes as written, and fails if the previous commit is reverted.

wchargin-branch: tarfile-limit-device-headers

ethanfurman reviewed (changes requested):
Looking good.  Nice tests.

Please add yourself to Misc/ACKS and we should be good to go.  (Sorry for not noticing that earlier.)

wchargin committed:
review: add wchargin to Misc/ACKS

(Per request of Ethan Furman.)

wchargin-branch: tarfile-limit-device-headers
